# Домашняя работа # 2

Всего баллов: 12 баллов (идет в ведомость максимум 11).

За неаккуратное оформление работа штрафуется максимум на 1 балл с шагом 0.25 балла (нерабочий код, неудаленные черновики и т.д.).

Импортируем библиотеки.


In [145]:
import pandas as pd
import matplotlib.pyplot as plt

Будем работать с [данными](https://www.ssa.gov/oact/babynames/limits.html) о самых популярных именах детей, родившихся в США в 1880-2020 годах.

Пример файла и легенда:

In [146]:
example = pd.read_csv('https://raw.githubusercontent.com/rogovich/Data/master/data/babies%20names/yob1880.txt',
                      names = ['name', 'sex', 'number'])
example.head()

name sex  number
0       Mary   F    7065
1       Anna   F    2604
2       Emma   F    2003
3  Elizabeth   F    1939
4     Minnie   F    1746

Данные:

* `name` - имя
* `sex` - пол (мужское или женское)
* `number` - количество детей с таким именем

Файл из примера содержит данные за 1880 год. Т.е. последняя строчка в примере читается так:
"В 1880-м году 1746 девочек назвали Minnie"


## Задание 1 (2 балла)

**🔥 Если у вас не получилось сделать это задание, то в следующем задании можно загрузить готовый файл.**

**🔥🔥 Данное задание засчитывается полностью, если реализован предложенный алгоритм. Другие решения могут быть засчитаны частично.**

По ссылке скачайте архив с файлами для каждого года наблюдений (1880-2020). Распакуйте файлы в папку `data`. Это не нужно делать с помощью Pyhton!

Положите этот блокнот в папку вместе с данными. В вашей папке должен быть блокнот и папка `data` с распакованным архивом.

В первую очередь нам нужно собрать все файлы в один датасет.

* Импортируйте библиотеку `os` — это библиотека для работы с операционной системой. С ее помощью мы сможем получить список названий всех 140 файлов.
* создайте [пустой датафрейм pandas](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) `babies` со следующими колонками: `'name', 'sex', 'number', 'year'`
* Если вы все сделали правильно с распаковкой файлов, то `sorted(os.listdir('data'))` даст вам список всех файлов в папке `data`, отсортированный от самого раннего года к самому позднему.
* Запустите цикл `for` по `sorted(os.listdir('data'))` — на каждой итерации во временну переменную будет сохраняться название файла. Для каждого названия файла сделайте следующее:

    * Проверьте, что в название файла входит строка `'yob'` (так мы защитимся от возможных ошибок, если в папке будут какие-то посторонние файлы, например, временные)
    * С помощью среза извлеките из названия файла год, превратите в целое число и сохраните в переменную `year`.
    * С помощью функции `pd.read_csv()` откройте файл как датафрейм и сохраните его в переменную с названием `temp`. Ваш путь к файлу должен начинаться с `'data/'`. Название файла можете приклеить к этой строке с помощью оператора `+`. В качестве дополнительного параметра укажите `names=['name', 'sex', 'number']`, чтобы добавить названия колонок.
    * В датафрейме `temp` создайте новую колонку `'year'` и запишите в нее год из переменной `year`. 
    * Таким образом мы получим датафрейм для, например, 1880-го года, в который мы к каждому наблюдению добавим значение года `1880` в новую колонку.
    * Теперь нам нужно с помощью метода датафрейма `.append()` добавить только считанный датафрейм `temp` в конец датафрейма `babies`. Вызовите метод от `babies` и передайте `temp` в качестве аргумента. Добавьте параметр `ignore_index=True`, чтобы индекс в итоговом датафрейме был сквозной. Не забудьте перезаписать объект `babies`! `.append()` не изменяет исходный объект.
    * Этот цикл будет работать не быстро. Давайте сделаем своеобразный 'progress bar', чтобы знать, что все хорошо. Сделайте проверку условия: если `year` кратен 10, то напечатай `<НАЗВАНИЕ ФАЙЛА> is done`. Т.е. Python будет нам выводить сообщение, когда обработан каждый 10-й файл, а мы не будем волноваться, что код долго работает.

In [154]:
# импортируйте библиотеку os

# создайте пустой датафрейм `babies` с колонками 'name', 'sex', 'number', 'year'

# запустите цикл for по объекту sorted(os.listdir('data'))

# сделайте проверку, что строка 'yob' является частью названия файла
        # извлеките из названия год, превратите в целое число и сохраните в `year`
        # считайте файл как датафрейм в `temp`, добавьте названия колонок `names=['name', 'sex', 'number']`
        # добавьте в `temp` новую колонку `year` и запишите в нее год из переменной `year`
        # с помощью .append() добавьте `temp` в конец `babies`, добавьте `ignore_index=True`. Не забудьте перезаписать `babies`!
    
        # добавьте проверку, что год в `year` кратен 10
            # если кратен, напечатайте строку '<НАЗВАНИЕ ФАЙЛА> is done'

В итоге у вас должен получиться датафрейм `babies` размерности `(2020863, 4)`. Если все правильно, то следующая ячейка при запуске выдаст `True`

In [141]:
babies_to_check = pd.read_csv('https://github.com/rogovich/2022_POL_Data_Analysis_in_Python/blob/main/@HW/HW2/all_babies_1880_2020.csv?raw=true')
print(len(babies.compare(babies_to_check)) == 0)

True


А `.tail()` датафрейма будет выглядеть вот так

In [49]:
babies_to_check.tail()

name sex number  year
2020858  Zykell   M      5  2020
2020859   Zylus   M      5  2020
2020860  Zymari   M      5  2020
2020861     Zyn   M      5  2020
2020862   Zyran   M      5  2020

## Задание 2 (1 балл)
Исследуйте набор данных `babies`. Ответьте на вопросы, каждый вопрос 0.2 балла. Если в предыдущем задании у вас не получилось создать нужный датафрейм, то раскомментируйте код ниже и запустите его.

In [142]:
# babies = pd.read_csv('https://github.com/rogovich/2022_POL_Data_Analysis_in_Python/blob/main/@HW/HW2/all_babies_1880_2020.csv?raw=true')


1) Какие годы входят в наш набор данных? Напечатайте все уникальные значения.

In [147]:
# ( ͡° ͜ʖ ͡°)づ ━━ ✫・*。

2) Какое имя в датасете находится по индексом 121? Выведите только имя.

In [148]:
# ( ͡° ͜ʖ ͡°)づ ━━ ✫・*。

3) Cколько всего родилось детей по имени 'Aaron' за все время?



In [149]:
# ( ͡° ͜ʖ ͡°)づ ━━ ✫・*。

4) Cколько мальчиков родилось в 2010?

In [150]:
# ( ͡° ͜ʖ ͡°)づ ━━ ✫・*。

5) Сколько в датасете девочек по имени John?

In [151]:
# ( ͡° ͜ʖ ͡°)づ ━━ ✫・*。

## Задание 3 (2.5 балла)

Каждый пункт 0.5 балла.

1. Отфильтруйте набор данных `babies` по полу, сгруппируйте по годам и для каждого года найдите суммарное количество рожденных детей. Сохраните результаты в два новых датафрейма: `babies_girls` и `babies_boys`. Т.е. должно получиться два датафрейма (девочки и мальчики), где каждому году соответствует общее количество детей, рожденных в этом году.
2. Создайте рисунок `matplotlib` с 3 графиками один под другим. Добавьте параметр `figsize=(7, 15)`. Постройте линейные графики. Первый график должен показывать тренд рождаемости для девочек (фиолетовый цвет), второй - для мальчиков (зеленый цвет), третий объединять их все вместе (с теми же цветами, что в индивидуальных графиках). 
3. Верхняя и правая границы графиков должны быть невидимы, к каждому графику должен быть заголовок, третий график должен содержать легенду, шкалы графиков должны быть подписаны — 'год' и 'количество новорожденных (млн)'
4. Для шкалы y, показывающей количество детей, рожденных в этот год, должны быть установлены лимиты (0, 2.30e+06), чтобы она была одинакова на обоих графиках.
5. В новой ячейке markdown под графиками напишите краткий вывод, отличаются ли тренды рождаемости для мальчиков и девочек.

In [152]:
# ( ͡° ͜ʖ ͡°)づ ━━ ✫・*。

## Задание 4 (2.5 балла)

Каждый пункт 0.5 балла

1. В оригинальном датафрейме `babies` создайте новую колонку - первая буква имени `first_letter`.
2. Выберете год из датасета. Сгруппируйте датасет так, чтобы его индексом были первые буквы, а колонкой - количество детей, рожденных в этот год, с именами, начинающимися на эти буквы. Сохраните два новых датафрейма для 1880 и 2020 года (например, `babies_first_1880`).
3. Создайте рисунок `matplotlib` с 2 графиками один под другим. Добавьте параметр `figsize=(7, 10)`. Постройте столбчатую диаграмму для каждого года в отдельном графике.
4. Верхняя и правая границы графиков должны быть невидимы, каждый график быть с заголовком, шкалы графиков должны быть подписаны.
5. В новой ячейке markdown под графиками кратко опишите как изменились тренды в именах, если смотреть на первые буквы.

In [ ]:
# ( ͡° ͜ʖ ͡°)づ ━━ ✫・*。

## Задание 5 (4 балла)

Каждый пункт 0.8 балла

1. Создайте пустой список `popular_names`. Найдите самые полулярные женские и мужские имена в 1880 году и 2020 году (всего 4 имени) и добавьте из в список `popular_names`.
2. Создайте рисунок matplotlib с одним горизонтальным графиком, размер установите (10, 5). В одной оси координат постройте линейный графики для каждого популярного за все время (это можно сделать в цикле со списком `popular_names`). На каждом шаге цикла сгруппируйте отфильтрованный датасет для этого имени по году и найдите общество количество новорожденных и постройте график для этих данных. 
3. Верхняя и правая границы графиков должны быть невидимы, каждый график быть с заголовком, шкалы графиков должны быть подписаны, у графиков должна быть легенда (передайте методу `.legend()` `popular_names` в качестве параметра. Для каждого из графиков выберите нестандартный цвет в палитре hex-color.
4. Адаптируйте и добавьте код к вашему графику, чтобы каждые 20 лет начиная с 1880-го года по 2020-й были отмечены тонкой серой линией (тут вам поможет range() c шагом. 

```
for year in [1880, 1890]:
    ax.axvline(x=year, color='blue', lw=0.25)
```
5. В новой ячейке markdown под графиками кратко опишите как тренды четырех имен.

In [153]:
# ( ͡° ͜ʖ ͡°)づ ━━ ✫・*。